backtrace-library/src/main/java/backtraceio/library/breadcrumbs/BacktraceBreadcrumbs.java

```diff
@@ -0,0 +1,342 @@
+     */
+    public boolean addBreadcrumb(String message, Map<String, Object> attributes, BacktraceBreadcrumbType type) {
+        return addBreadcrumb(message, attributes, type, BacktraceBreadcrumbLevel.INFO);
+    }
+
+    /**
+     * Add a breadcrumb of the desired level and type with the provided message string and attributes
+     * @param message
+     * @param attributes
+     * @param type
+     * @param level
+     * @return true if the breadcrumb was successfully added
+     */
+    public boolean addBreadcrumb(String message, Map<String, Object> attributes, BacktraceBreadcrumbType type, BacktraceBreadcrumbLevel level) {
+        if (this.isBreadcrumbsEnabled() && backtraceBreadcrumbsLogger != null) {
+            return backtraceBreadcrumbsLogger.addBreadcrumb(message, attributes, type, level);
+        }
+        return false;
+    }
+
+    /**
+     * If Breadcrumbs is currently enabled, process the BacktraceReport for sending the Breadcrumb logs
+     * @param backtraceReport
+     */
+    public void processBacktraceReport(BacktraceReport backtraceReport) {
+        if (this.isBreadcrumbsEnabled()) {
+            backtraceReport.attachmentPaths.add(BacktraceBreadcrumbs.getBreadcrumbLogPath(context));
+
+            long lastBreadcrumbId = this.getCurrentBreadcrumbId();
+            backtraceReport.attributes.put("breadcrumbs.lastId", lastBreadcrumbId);
+        }
+    }
+
+    /**
+     * Create a breadcrumb which reflects the current breadcrumb configuration
+     * @return true if the breadcrumb was successfully added
+     */
+    private boolean addConfigurationBreadcrumb()
+    {
+        if (backtraceBreadcrumbsLogger == null) {
+            return false;
+        }
+
+        Map<String, Object> attributes = new HashMap<String, Object>();
+
+        for (BacktraceBreadcrumbType enabledType : BacktraceBreadcrumbType.values())
+        {
+            if (enabledType == BacktraceBreadcrumbType.CONFIGURATION) {
+                attributes.put(enabledType.toString(), "enabled");
+            }
+
+            String state = (enabledBreadcrumbTypes != null &&
+                    enabledBreadcrumbTypes.contains(enabledType)) ? "enabled" : "disabled";
+
+            attributes.put(enabledType.toString(), state);
+        }
+
+        return backtraceBreadcrumbsLogger.addBreadcrumb("Breadcrumbs configuration",
+                attributes,
+                BacktraceBreadcrumbType.CONFIGURATION,
+                BacktraceBreadcrumbLevel.INFO);
+    }
+
+    private void handleExistingBreadcrumbLogFile(Context context, BacktraceBase breadcrumbLogSender) {
+        if (context == null || breadcrumbLogSender == null) {
+            return;
+        }
+
+        final File breadcrumbLogFile = new File(BacktraceBreadcrumbs.getBreadcrumbLogPath(context));
+        final File breadcrumbTempLogFile = new File(this.getBreadcrumbLogDirectory() + "/bt-stalecrumbs");
+        try {
+            if (breadcrumbLogFile.exists()) {
```

Comment:
After moving this function to Breadcrumbs class we lost conversation history. 

From user perspective: 
* it won't be suprising that we will send a report with breadcrumbs on application startup without any good reason?

From reporting flow perspective:
* How often breadcrumbs log that we will send from this method will have useful information for user? 

I think we're talking about the really rare situation that our native integration already handles for us. On the other hand sending a report on the application startup with just an old breadcrumb file might be confussing for people - I don't see a value for user from this report. User also won't be able to process all those reports - we will treat them the same and put every of them under the same fingerprint. Each report will have different breadcrumbs and I think after hours of manual work on trying to find useful report with old breadcrumbs we will recognize we didn't find anything useful. 

What do you think?

Reply:
I try to do a few things to make this feature more useful:

- I add a shutdownHook to `BacktraceQueueFileHelper` (I don't know how to link code from my PR) so that if we exit "cleanly" we will delete the breadcrumbs log file (i.e: user exits application properly). If the user even force closes the app it won't be a clean shutdown and the breadcrumbs log file will persist. That can definitely be a bunch of distracting noise, but maybe it can be helpful to have a breadcrumb trail leading up to a fatal crash as well. We could possibly tighten this up more by beefing up the shutdown/app killed detection logic so we delete the breadcrumbs log file in more cases.  
- The main intention here is if we want breadcrumbs for fatal crashes too. If we want breadcrumbs for fatal crashes, we would need some way to keep breadcrumbs from the session until next time when we send the report "hey you had a fatal crash last time." But you bring up a good point...I should figure out how Crashpad handles this case and attach our logic to there.

Reply:
For now we will overwrite the old Breadcrumb log always. You are right, this is not the right way to handle this.